File: packages/js-dpp/lib/stateTransition/fee/operations/ReadOperation.js
const AbstractOperation = require('./AbstractOperation');

const {
  READ_BASE_PROCESSING_COST,
  PROCESSING_CREDIT_PER_BYTE,
} = require('../constants');

class ReadOperation extends AbstractOperation {
  /**
   * @param {number} valueSize
   */
  constructor(valueSize) {
    super();

    this.valueSize = valueSize;
  }

  /**
   * Get CPU cost of the operation
   *
   * @returns {number}
   */
  getProcessingCost() {
    return READ_BASE_PROCESSING_COST + this.valueSize * PROCESSING_CREDIT_PER_BYTE;
  }

  /**
   * Get storage cost of the operation
   *
   * @returns {number}
   */
  getStorageCost() {
    return 0;
  }

  /**
   * @return {{valueSize: number, type: string}}
   */
  toJSON() {
    return {
      type: 'read',
      valueSize: this.valueSize,
    };
  }

  /**
   * @param {{type: string, valueSize: number}} json
   * @return {ReadOperation}
   */
  static fromJSON(json) {
    return new ReadOperation(json.valueSize);
  }
}

module.exports = ReadOperation;

File: packages/rs-dpp/src/state_transition/state_transition_fee/operations/read_operation.rs
use serde::{Deserialize, Serialize};

use super::OperationLike;

use crate::fee::default_costs::constants::{PROCESSING_CREDIT_PER_BYTE, READ_BASE_PROCESSING_COST};
use crate::fee::Credits;
use crate::NonConsensusError;

#[derive(Default, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ReadOperation {
    pub value_size: Credits,
}

impl ReadOperation {
    pub fn new(value_size: u64) -> Self {
        Self { value_size }
    }
}

impl OperationLike for ReadOperation {
    fn get_processing_cost(&self) -> Result<Credits, NonConsensusError> {
        let value_byte_processing_cost = self
            .value_size
            .checked_mul(PROCESSING_CREDIT_PER_BYTE)
            .ok_or(NonConsensusError::Overflow(
                "value processing cost is too big",
            ))?;

        READ_BASE_PROCESSING_COST
            .checked_add(value_byte_processing_cost)
            .ok_or(NonConsensusError::Overflow(
                "can't add read base processing cost",
            ))
    }

    fn get_storage_cost(&self) -> Result<Credits, NonConsensusError> {
        Ok(0)
    }

    fn get_refunds(&self) -> Option<&Vec<Refunds>> {
        None
    }
}
